### PR TITLE
Encrypt region boundary keys, Part 1 - helper functions

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Signed-off-by: Yi Wu <yiwu@pingcap.com>

### What problem does this PR solve?

This is part 1 for adding TDE support to PD. https://github.com/pingcap/tidb/issues/18262 It contains helper methods for encryption.

### What is changed and how it works?

This PR adds utility types and functions:
* encrypt/decrypt data using aes-ctr (for encrypting region boundary keys) and aes-gcm (for encrypting data encryption keys)
* helper methods to encrypt/decrypt region boundary keys in-place
* wrapper of master key, and helper method to read a master key (must be a 256 bit encryption key, stored as hex-string) from a local file

### Check List

Tests

- Unit test

Related changes

- depends on https://github.com/pingcap/kvproto/pull/666

### Release note

* No release note

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -54,3 +55,5 @@ require (
 )
 
 replace go.etcd.io/bbolt => go.etcd.io/bbolt v1.3.5
+
+replace github.com/pingcap/kvproto => github.com/yiwu-arbug/kvproto v0.0.0-20200908220833-164267066987
```
Comment (yiwu-arbug) [inline go.mod:59]: will revert once https://github.com/pingcap/kvproto/pull/666 is merged.

File: go.mod [hunk 2]
```diff
@@ -34,6 +34,7 @@ require (
 	github.com/pingcap/kvproto v0.0.0-20200827082727-23dedec2339b
 	github.com/pingcap/log v0.0.0-20200511115504-543df19646ad
 	github.com/pingcap/sysutil v0.0.0-20200715082929-4c47bcac246a
+	github.com/pkg/errors v0.9.1
```
Comment (HunDunDM) [inline go.mod:37]: Please use `github.com/pingcap/errors`, and add relevant information in `pkg/errs/errno` according to the needs of the log.

File: pkg/encryption/crypter.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+func KeyLength(method encryptionpb.EncryptionMethod) int {
+	switch method {
+	case encryptionpb.EncryptionMethod_AES128_CTR:
+		return 16
+	case encryptionpb.EncryptionMethod_AES192_CTR:
+		return 24
+	case encryptionpb.EncryptionMethod_AES256_CTR:
+		return 32
+	default:
+		panic("unsupported encryption method")
```
Comment (Yisaer) [inline pkg/encryption/crypter.go:62]: I tend to throw an error here. Why use panic ?

Comment (HunDunDM) [reply pkg/encryption/crypter.go:62]: I think it's better to use `log.Fatal`

File: pkg/encryption/region_crypter.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	if region == nil {
+		return nil
+	}
+	if region.EncryptionMeta != nil && region.EncryptionMeta.IsEncrypted {
+		return errors.Errorf("region already encrypted, region id = %d", region.Id)
+	}
+	if keyManager == nil {
+		// encryption is not enabled.
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply pkg/encryption/region_crypter.go:50]: `keyManager == nil` always mean encryption is disabled, not encryption keys are missing or other problems. Or do you suggest any pattern I should follow?

File: pkg/encryption/master_key_test.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+			},
+		},
+	}
+	_, err := NewMasterKey(config)
+	c.Assert(err, Not(IsNil))
+}
+
+func (s *testMasterKeySuite) TestNewFileMasterKeyMissingFile(c *C) {
+	dir, err := ioutil.TempDir("", "test_key_files")
+	c.Assert(err, IsNil)
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply pkg/encryption/master_key_test.go:102]: `ioutil.TempDir` will generate random dirname by appending a random number to the file name, according to the docs, though I didn't really test it. https://golang.org/pkg/io/ioutil/#TempDir

**Comment by yiwu-arbug** [comment]
comments addressed or answered. PTAL.

File: pkg/encryption/master_key_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/encryption/master_key_test.go:102]: Thanks for pointing it out.

**Review by Yisaer — APPROVED** [review]
LGTM

File: pkg/encryption/crypter.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+		return
+	}
+	keyIDBuf := make([]byte, 8)
+	n, err := io.ReadFull(rand.Reader, keyIDBuf)
+	if err != nil {
+		err = errs.ErrEncryptionNewDataKey.Wrap(err).GenWithStack(
+			"fail to generate data key id")
+		return
+	}
+	if n != 8 {
```
Comment (HunDunDM) [inline pkg/encryption/crypter.go:114]: I think we need `const uint64Size = 8`

File: pkg/encryption/master_key.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+const (
+	// Master key is of fixed 256 bits (32 bytes).
+	masterKeyLength = 32 // in bytes
+)
+
+// MasterKey is used to encrypt and decrypt encryption metadata (i.e. data encryption keys).
+type MasterKey struct {
+	// Master key config. Used to compare if two master key is the same.
+	Config *encryptionpb.MasterKey
```
Comment (HunDunDM) [inline pkg/encryption/master_key.go:34]: Do we need to rename it to `MasterKeyConfig` in `kvproto`?

File: pkg/encryption/master_key.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+		key, err := newMasterKeyFromFile(file)
+		if err != nil {
+			return nil, err
+		}
+		return &MasterKey{
+			Config: config,
+			key:    key,
+		}, nil
+	}
+	return nil, errors.New("unrecognized master key type")
```
Comment (HunDunDM) [inline pkg/encryption/master_key.go:62]: Doesn't it support `MasterKeyKms`?

File: pkg/encryption/crypter.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+// You may obtain a copy of the License at
+//
+//	   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encryption
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply pkg/encryption/crypter.go:14]: security currently include TLS and encryption (TDE). I think `encryption` is better.

File: pkg/encryption/crypter.go (hunk 6 above)
Comment (yiwu-arbug) [reply pkg/encryption/crypter.go:114]: Is there a Go lang constant that I can use?

File: pkg/encryption/master_key.go (hunk 8 above)
Comment (yiwu-arbug) [reply pkg/encryption/master_key.go:62]: Will add in later PR.

File: pkg/encryption/master_key.go (hunk 7 above)
Comment (yiwu-arbug) [reply pkg/encryption/master_key.go:34]: I think both are okay?

File: pkg/encryption/crypter.go (hunk 6 above)
Comment (HunDunDM) [reply pkg/encryption/crypter.go:114]: `unsafe.Sizeof(uint64(0))`

Comment (yiwu-arbug) [reply pkg/encryption/crypter.go:114]: addressed in #2976

File: pkg/encryption/crypter.go [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+	return iv, nil
+}
+
+// NewIvCtr randomly generate an IV for CTR mode.
+func NewIvCtr() (IvCtr, error) {
+	return newIV(ivLengthCTR)
+}
+
+// NewIvGcm randomly generate an IV for GCM mode.
+func NewIvGcm() (IvGcm, error) {
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply pkg/encryption/crypter.go:95]: addressed in #2976

File: pkg/encryption/crypter.go [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+	}
+	if n != ivLength {
+		return nil, errs.ErrEncryptionGenerateIV.GenWithStack(
+			"iv length exepcted %d vs actual %d", ivLength, n)
+	}
+	return iv, nil
+}
+
+// NewIvCtr randomly generate an IV for CTR mode.
+func NewIvCtr() (IvCtr, error) {
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply pkg/encryption/crypter.go:90]: addressed in #2976

File: pkg/encryption/crypter.go [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+		}
+		return 0, errs.ErrEncryptionInvalidMethod.GenWithStackByArgs(int32(method))
+	}
+}
+
+// IvCtr represent IV bytes for CTR mode.
+type IvCtr []byte
+
+// IvGcm represent IV bytes for GCM mode.
+type IvGcm []byte
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply pkg/encryption/crypter.go:74]: addressed in #2976

File: pkg/encryption/crypter.go [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+		name, ok := encryptionpb.EncryptionMethod_name[int32(method)]
+		if ok {
+			return 0, errs.ErrEncryptionInvalidMethod.GenWithStackByArgs(name)
+		}
+		return 0, errs.ErrEncryptionInvalidMethod.GenWithStackByArgs(int32(method))
+	}
+}
+
+// IvCtr represent IV bytes for CTR mode.
+type IvCtr []byte
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply pkg/encryption/crypter.go:71]: addressed in #2976

**Pull request merged**